src/jit/codegenarm64.cpp

```diff
@@ -438,107 +431,114 @@ void CodeGen::genSaveCalleeSavedRegistersHelp(regMaskTP   regsToSaveMask,
 
     int spOffset = lowestCalleeSavedOffset; // this is the offset *after* we change SP.
 
-    if (maskSaveRegsInt != RBM_NONE)
-    {
-        // Save the integer registers
-
-        unsigned intRegsToSaveCount = genCountBits(maskSaveRegsInt);
-        bool lastSavedWasPair = false;
+    unsigned intRegsToSaveCount = genCountBits(maskSaveRegsInt);
+    unsigned floatRegsToSaveCount = genCountBits(maskSaveRegsFloat);
+    bool isPairSave = false;
+#ifdef DEBUG
+    bool isRegsToSaveCountOdd = ((intRegsToSaveCount + floatRegsToSaveCount) % 2 != 0);
+#endif
 
-        while (maskSaveRegsInt != RBM_NONE)
-        {
-            regMaskTP reg1Mask = genFindLowestBit(maskSaveRegsInt);
-            regNumber reg1 = genRegNumFromMask(reg1Mask);
-            maskSaveRegsInt &= ~reg1Mask;
+    // Save the integer registers
 
-            if (intRegsToSaveCount >= 2)
-            {
-                // We can use a STP instruction.
+    bool lastSavedWasPair = false;
 
-                regMaskTP reg2Mask = genFindLowestBit(maskSaveRegsInt);
-                regNumber reg2 = genRegNumFromMask(reg2Mask);
-                assert((reg2 == REG_NEXT(reg1)) || (reg2 == REG_LR));
-                maskSaveRegsInt &= ~reg2Mask;
+    while (maskSaveRegsInt != RBM_NONE)
+    {
+        // If this is the first store that needs to change SP (spDelta != 0),
+        // then the offset must be 8 to account for alignment for the odd count
+        // or it must be 0 for the even count.
+        assert((spDelta == 0) || (isRegsToSaveCountOdd && spOffset == REGSIZE_BYTES)
+            || (!isRegsToSaveCountOdd && spOffset == 0));
 
-                genPrologSaveRegPair(reg1, reg2, spOffset, spDelta, lastSavedWasPair, REG_IP0, nullptr);
+        isPairSave = (intRegsToSaveCount >= 2);
+        regMaskTP reg1Mask = genFindLowestBit(maskSaveRegsInt);
+        regNumber reg1 = genRegNumFromMask(reg1Mask);
+        maskSaveRegsInt &= ~reg1Mask;
+        intRegsToSaveCount -= 1;
 
-                // TODO-ARM64-CQ: this code works in the prolog, but it's a bit weird to think about "next" when generating this epilog, to
-                // get the codes to match. Turn this off until that is better understood.
-                // lastSavedWasPair = true;
+        if (isPairSave)
+        {
+            // We can use a STP instruction.
 
-                intRegsToSaveCount -= 2;
-                spOffset += 2 * REGSIZE_BYTES;
-            }
-            else
-            {
-                // No register pair; we use a STR instruction.
+            regMaskTP reg2Mask = genFindLowestBit(maskSaveRegsInt);
+            regNumber reg2 = genRegNumFromMask(reg2Mask);
+            assert((reg2 == REG_NEXT(reg1)) || (reg2 == REG_LR));
```

Comment:
There is no requirement that reg2 == REG_NEXT(reg1) in a stp instruction, why assert this?